fix(plugin-vue): respect __VUE_PROD_DEVTOOLS__ setting

oliverzy opened the pull request:
### Description

If user defines __VUE_PROD_DEVTOOLS__ to true, plugin-vue should respect this to make sure the generated code can be compatible with Vue devtools (inlineTemplate and filename) 

### Additional context

script setup with inlineTemplate as true is not compatible with Vue devtools since it doesn't expose any bindings.
Vue devtools relies on filename to infer the component name for the component using script setup.

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [X] Bug fix

### Before submitting the PR, please make sure you do the following

- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [X] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [X] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [X] Ideally, include relevant tests that fail without this PR but pass with it.

oliverzy committed:
fix(plugin-vue): respect __VUE_PROD_DEVTOOLS__ setting

Commit:
Merge branch 'main' into respect-prod-dev-tools-setting

haoqunjiang reviewed (changes requested):
Please don't commit style-only changes on files irrelevant to the PR.

oliverzy committed:
chore: merge

# Conflicts:
#	packages/playground/ssr-vue/example-external-component/ExampleExternalComponent.vue

oliverzy committed:
chore: reset

oliverzy commented:
OK, it's done.

sodatea commented on packages/plugin-vue/src/main.ts:
```diff
@@ -58,6 +58,7 @@ export async function transformMain(
   // inlined template cannot be individually hot updated.
   const useInlineTemplate =
     !devServer &&
+    !devToolsEnabled &&
```
Seems exposing `__filename` is enough for the devtool to display the component name, why do you want to disable `inlineTemplate`?

oliverzy replied:
Because for component with script setup, inlineTemplate option doesn't expose any setup binding. Vue devtools relies on this to display component setup binding correctly.

sodatea replied:
Got it.
I have discussed this with Evan today.
As the `inlineTemplate` has an impact on the app performance, we believe that it shouldn't be disabled.
We need to find another way to fix this issue. Maybe in another package.

sodatea commented on packages/plugin-vue/src/main.ts:
```diff
@@ -108,7 +109,7 @@ export async function transformMain(
-  if (devServer && !isProduction) {
+  if (devToolsEnabled || (devServer && !isProduction)) {
     // expose filename during serve for devtools to pickup
     attachedProps.push([`__file`, JSON.stringify(filename)])
```
Due to safety concerns, the full path should not be used in production. Please use `path.basename(filename)`.
The development bundle should still use the full path so that users can open the file directly from devtools.

oliverzy replied:
👌

derrickb replied:
Hi, I've just started using `@vitejs/plugin-vue2` in my rollup project, and was surprised to see the full filepath to my modules included in the generated output (as `__component__.options.__file`), and have not found a way to disable this. Can you advise on how to remove these filepaths?

Commit:
fix: doesn't expose full file path when production

haoqunjiang reviewed (changes requested):
Please remove the changes related to `inlineTemplate`.
We need to open an issue in the `vue-next` repo to provide the ability required by Vue DevTools, rather than work around the issue in the Vite plugin.

The `__file` fix looks good to me.

flxsource commented:
> Please remove the changes related to `inlineTemplate`. We need to open an issue in the `vue-next` repo to provide the ability required by Vue DevTools, rather than work around the issue in the Vite plugin.
> 
> The `__file` fix looks good to me.

Has the `vue-next` (now `vuejs/core`?) issue been raised? I haven't been able to find anything relevant, but want to make sure I'm not duplicating the issue.

sodatea commented:
Raised in https://github.com/vuejs/core/issues/5496

haoqunjiang committed:
fix: apply review suggestions, always use `inlineTemplate` in production

haoqunjiang committed:
chore: merge branch 'main' into respect-prod-dev-tools-setting

patak-dev commented:
@sodatea is this one ready to merge?

sodatea commented:
@patak-dev Yeah, I forgot to approve it after pushing…
